File: pipeline/models/feats.py
"""Pydantic model for feats.json."""

from __future__ import annotations

from typing import Literal

from .common import LenientModel, StrictModel


class FeatsMetadata(StrictModel):
    """Metadata block at the top of feats.json."""

    description: str
    version: str
    sectionsComplete: list[str]
    sectionsPending: list[str]


FeatCategory = Literal[
    "general",
    "racial",
    "supplementary",
    "asset",
    "exaltedAsset",
    "hindrance",
]


class Feat(LenientModel):
    """A single feat, asset, or hindrance.

    Optional fields present on subsets of feats:
    - raceRestriction: racial feats limited to a specific race
    - exaltationRestriction: exalted assets limited to a specific exaltation
    - creationOnly: assets/hindrances only available at character creation
    - bonusXP: hindrances that grant bonus XP
    - subOptions: feats with nested choice options (e.g. Wizard Tradition)
    """

    id: str
    name: str
    category: FeatCategory
    effect: str
    details: str
    multipleAllowed: bool
    groups: list[str] | None
    prerequisites: str | None
    raceRestriction: str | None = None
    exaltationRestriction: str | None = None
    creationOnly: bool | None = None
    bonusXP: int | None = None
    subOptions: list[dict] | dict[str, str] | None = None


class FeatsFile(StrictModel):
    """Top-level structure of feats.json."""

    metadata: FeatsMetadata
    feats: list[Feat]

File: src/lib/dtd/schemas/feats.ts
/**
 * Zod schemas for feat data.
 *
 * Mirrors pipeline/models/feats.py.
 */

import { z } from "zod";

// ---------------------------------------------------------------------------
// Sub-schemas
// ---------------------------------------------------------------------------

export const FeatsMetadata = z
	.object({
		description: z.string(),
		version: z.string(),
		sectionsComplete: z.array(z.string()),
		sectionsPending: z.array(z.string()),
	})
	.strict();
export type FeatsMetadata = z.infer<typeof FeatsMetadata>;

export const FeatCategory = z.enum(["general", "racial", "supplementary", "asset", "exaltedAsset", "hindrance"]);
export type FeatCategory = z.infer<typeof FeatCategory>;

// ---------------------------------------------------------------------------
// Top-level schemas
// ---------------------------------------------------------------------------

export const Feat = z
	.object({
		id: z.string(),
		name: z.string(),
		category: FeatCategory,
		effect: z.string(),
		details: z.string(),
		multipleAllowed: z.boolean(),
		groups: z.array(z.string()).nullable(),
		prerequisites: z.string().nullable(),
		raceRestriction: z.string().nullish(),
		exaltationRestriction: z.string().nullish(),
		creationOnly: z.boolean().nullish(),
		bonusXP: z.number().int().nullish(),
		subOptions: z.union([z.array(z.record(z.string(), z.unknown())), z.record(z.string(), z.string())]).nullish(),
	})
	.passthrough();
export type Feat = z.infer<typeof Feat>;

export const FeatsFile = z
	.object({
		metadata: FeatsMetadata,
		feats: z.array(Feat),
	})
	.strict();
export type FeatsFile = z.infer<typeof FeatsFile>;
